[git-webkit] Identify all unpicked pickable commits

JonWBedard opened the pull request:
#### 13e85ed5ecaed35f43d4db23a0d81220f7042022
<pre>
[git-webkit] Identify all unpicked pickable commits
<a href="https://bugs.webkit.org/show_bug.cgi?id=246542">https://bugs.webkit.org/show_bug.cgi?id=246542</a>
rdar://97399813

Reviewed by Dewei Zhu and Elliott Williams.

* Tools/Scripts/libraries/webkitscmpy/setup.py: Bump version.
* Tools/Scripts/libraries/webkitscmpy/setup.py: Bump version, add rapidfuzz dependency.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/__init__.py: Ditto.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pickable.py:
(fuzzy_filter): Use rapidfuzz if it&apos;s available, otherwise use regexes.
(Pickable): Specify default patern categories.
(Pickable.parser): Add --into, --exclude, --scope, --filter and --by arguments.
(Pickable.pickable): Accept commits story, use global exclude list and named filters.
(Pickable.main): Construct a commits story to pass to pickable, use --filter argyuments
to filter commits based on commit message, use --by to filter commits based on author and
pass --scope into commit list to scope pickable query to a specific directory path.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/pickable_unittest.py:
(TestPickable.test_none):
(TestPickable.test_main_no_into):
(TestPickable.test_main):
(TestPickable.test_scope):
(TestPickable.test_main_git_svn):
(TestPickable.test_branch_none):
(TestPickable.test_branch_include_versioning):

Canonical link: <a href="https://commits.webkit.org/255979@main">https://commits.webkit.org/255979@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1efa89c8b6f265c0eea965d553a16b10ac9bdd43

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/94236 "1 style error") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/77/builds/3424 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/25285 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/8/builds/103914 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/164192 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [  ~~🧪 bindings~~](https://ews-build.webkit.org/#/builders/11/builds/98233 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/3453 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/31639 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/86588 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/99935 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/99892 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/2478 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/80644 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/86588 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/97827 "Passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/84397 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/25285 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/86588 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/38021 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/25285 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/20/builds/93239 "Passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/35906 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/25285 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/39785 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/80644 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/1947 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/41729 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/25285 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

dewei-zhu commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pickable.py:
```diff
@@ -20,22 +20,62 @@
+            fuzzy_filter(r'is an almost constant failure'),
+            fuzzy_filter('are constant failures'),
+            fuzzy_filter('tests consistently failing'),
+        ], 'Build-fix': [
```
Nit: new line for `'Build-fix'`

dewei-zhu commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pickable.py:
```diff
@@ -20,22 +20,62 @@
+        ], 'Build-fix': [
+            re.compile(r'Apply build[ -]fix'),
+            re.compile(r'Unreviewed build[ -]fix'),
+        ], 'Apply-patch': [re.compile(r'Apply patch')],
```
Ditto.

dewei-zhu commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pickable.py:
```diff
@@ -20,22 +20,62 @@
+            re.compile(r'Apply build[ -]fix'),
+            re.compile(r'Unreviewed build[ -]fix'),
+        ], 'Apply-patch': [re.compile(r'Apply patch')],
+        'None': [],
```
Is string `'None'` a valid pattern?

JonWBedard replied:
This is how you disable all filters. By default, named filters in `DEFAULT_PATTERNS` are all enabled. A user needs to pass a filter to disable the default ones, if you pass `--exclude=None`, you'll end up with no filters.

dewei-zhu commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pickable.py:
```diff
@@ -50,48 +90,102 @@ def parser(cls, parser, loggers=None):
 
+        if excluded is None:
+            excluded = cls.DEFAULT_PATTERNS
+        excluded = list(itertools.chain(
+            *[cls.PATTERNS[arg] if arg in cls.PATTERNS else [re.compile(arg)] for arg in excluded]
+        ))
```
What if we do ```
excluded = list(itertools.chain(
             *[cls.PATTERNS[arg].search if arg in cls.PATTERNS else [re.compile(arg).search] for arg in excluded]
         ))```

JonWBedard replied:
Not all of the patterns are regexes, some are `fuzzy_filter`s, which is why we check if the patterns are callable on line 151.

dewei-zhu commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pickable.py:
```diff
@@ -50,48 +90,102 @@ def parser(cls, parser, loggers=None):
             title = commit.message.splitlines()[0]
             all_commits[str(commit)] = commit
-            if search('Cherry-pick', title) or search('Versioning', title):
+            if any([ex(title) if callable(ex) else ex.search(title) for ex in excluded]):
```
If we construct excluded the way mentioned above, this can just be `if any(ex(title) for ex in excluded)`

dewei-zhu reviewed (approved):
r=me

emw-apple commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pickable.py:
```diff
@@ -20,22 +20,64 @@
+            fuzzy_filter(r'is a constant failure'),
+            fuzzy_filter(r'is an almost constant failure'),
+            fuzzy_filter('are constant failures'),
+            fuzzy_filter('tests consistently failing'),
```
I'm concerned these will filter out non-gardening commits, like https://github.com/WebKit/WebKit/commit/debb098c0323383b53f8a07901c55ee4762fcb20.

AFAICT, it is normal for us to file a bug including the phrase "is a constant failure", and it's also normal for an engineer to use that bug title when committing their fix.

JonWBedard replied:
This is  good point.

I think we should add an extra filter on these commits, but I also think that filter should be in a different PR, because it involves something a bit out of scope of the goals of this change. I'm inclined to only consider a change a gardening change if the change itself matches these conditions and also only modifies files in `LayoutTests`...this feels like a bit of a cheat, because it's not necessarily applicable outside of WebKit, and it requires an extra `git` operation for each implicated commit, but also, I think it would fix this problem for good.

How would you feel if we removed:
```
fuzzy_filter(r'is a constant failure'),
fuzzy_filter(r'is an almost constant failure'),
fuzzy_filter('are constant failures'),
fuzzy_filter('tests consistently failing'),
```
and revisited those in a follow-up PR?

emw-apple replied:
That makes sense to me.

emw-apple commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pickable.py:
```diff
@@ -20,22 +20,64 @@
+    ]
+    PATTERNS = {
+        'Cherry-pick': [re.compile(r'[Cc]herry[- ][Pp]ick')],
+        'Versioning': [re.compile(r'[Vv]ersioning')],
```
I think these might be too broad to be allowed to match anywhere in the commit message first line.

Examples of would-be false negatives for "cherry-pick": 4f144420a531751b9d7f43a741713ae5ca330b24 8de977259722
and for "versioning": 690a2333ae979af678f1d910cf455a129757f2ff f3e5aebe67ea0146cb14da9cd46e5e5db9410a10

JonWBedard replied:
The `cherry-pick` example is disabled by default in favor of a more narrow one.

Let me see if I can narrow "versioning" any further....

JonWBedard replied:
We can! Narrowed the `Versioning` check, along with a few others, mostly by tying them to the start of strings.

emw-apple reviewed (approved):
I'm still nervous about hiding commits based on magic words (see additional comment). I'm not nervous enough to r-, I just worry that this method isn't going to be reliable for integrators to trust 100% of the time.

JonWBedard committed:
[git-webkit] Identify all unpicked pickable commits
https://bugs.webkit.org/show_bug.cgi?id=246542
rdar://97399813

Reviewed by Dewei Zhu and Elliott Williams.

* Tools/Scripts/libraries/webkitscmpy/setup.py: Bump version.
* Tools/Scripts/libraries/webkitscmpy/setup.py: Bump version, add rapidfuzz dependency.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/__init__.py: Ditto.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pickable.py:
(fuzzy_filter): Use rapidfuzz if it's available, otherwise use regexes.
(Pickable): Specify default patern categories.
(Pickable.parser): Add --into, --exclude, --scope, --filter and --by arguments.
(Pickable.pickable): Accept commits story, use global exclude list and named filters.
(Pickable.main): Construct a commits story to pass to pickable, use --filter argyuments
to filter commits based on commit message, use --by to filter commits based on author and
pass --scope into commit list to scope pickable query to a specific directory path.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/pickable_unittest.py:
(TestPickable.test_none):
(TestPickable.test_main_no_into):
(TestPickable.test_main):
(TestPickable.test_scope):
(TestPickable.test_main_git_svn):
(TestPickable.test_branch_none):
(TestPickable.test_branch_include_versioning):

Canonical link: https://commits.webkit.org/255979@main
